### PR TITLE
Bithound.metalsmith stylus.2.0.0

### PR DESCRIPTION
**Pull request by NubleX**
Back-up security.

### CODE REVIEW

**Comment by ghost** [comment]
could you provide a little more information as to what exact security flaw is fixed here? also, it would be nice if you could fix the merge conflict 👍

**Comment by NubleX** [comment]
I'm thinking long-term here since I'm building an OS to operate the central nerve system and artificial limbs.

**Comment by ghost** [comment]
but... how does this relate to this website?

i do see the security flaw though. goes back to [this](https://snyk.io/vuln/npm:minimatch:20160620). fair enough, now we only need to fix this merge conflict

**Comment by lpinca** [comment]
I have no strong opinions, -0.

**Comment by fhemberger** [comment]
I think it would be a good idea, but I guess it won't be added to the org (just like greenkeeper was removed). I think @phillipj might know the reason.

File: package.json [hunk 1]
```diff
@@ -49,16 +51,17 @@
     "metalsmith-metadata": "0.0.2",
     "metalsmith-permalinks": "0.4.0",
     "metalsmith-prism": "2.1.1",
-    "metalsmith-stylus": "1.0.0",
+    "metalsmith-stylus": "^2.0.0",
```
Comment (stevemao) [inline package.json:54]: I think these should be in another pr

File: package.json [hunk 2]
```diff
@@ -11,7 +11,9 @@
     "test": "npm run test:lint && npm run test:unit && npm run test:smoke",
     "test:lint": "standard",
     "test:unit": "tape tests/**/*.test.js | faucet",
-    "test:smoke": "tape tests/*.smoketest.js | faucet"
+    "test:smoke": "tape tests/*.smoketest.js | faucet",
+    "snyk-protect": "snyk protect",
+    "prepublish": "npm run snyk-protect"
```
Comment (fhemberger) [inline package.json:16]: Website is not published to npm, so it should be

``` diff
-    "snyk-protect": "snyk protect",
-    "prepublish": "npm run snyk-protect"
+    "postinstall": "snyk protect"
```

File: package.json (hunk 1 above)
Comment (fhemberger) [inline package.json:54]: +1 for "Update dependencies" PR

Comment (lpinca) [inline package.json:54]: `metalsmith-stylus` is already at version `2.0.0` on master.
Some other dependencies updated here are also already updated on master.

File: package.json (hunk 2 above)
Comment (MylesBorins) [inline package.json:16]: isn't prepublish always run after install?

Comment (ghost) [inline package.json:16]: @TheAlphaNerd not always, but most times: 
screen shot 2016-07-09 at 23 32 30

Comment (fhemberger) [inline package.json:16]: I think the intention is better understandable with 'postinstall', even if 'prepublish' might work as well.

**Comment by phillipj** [comment]
I wasn't aware of snyk, at first glance it seems like a good thing to have. Although I would prefer if it would be less obtrusive, especially thinking of the `.snyk` file.

> I think it would be a good idea, but I guess it won't be added to the org (just like greenkeeper was removed). I think @phillipj might know the reason.

You're correct about not being able to grant the snyk GH application access. We could get around that issue by running it via Travis. I did some tests locally and seems like running `$ snyk test` does the job.

image

How do we want to integrate this in our current flow? E.g. should we break the build or just view snyk issues as (spammy) warnings? Or maybe just add this badge to README.md [](https://snyk.io/test/github/nodejs/nodejs.org) (yes, that badge is for nodejs.org).

And seeing the [current snyk report](https://snyk.io/test/github/nodejs/nodejs.org), we're not able to fix all issues as some dependencies seem to have been abandoned. Might need to replace a couple of deps and possibly change some of our code.

**Comment by fhemberger** [comment]
@phillipj `snyk protect` applies patches locally for those issues found, as long as no official fix/update is provided.

**Comment by phillipj** [comment]
@fhemberger sounds like blackmagic. What does that mean in detail? Got any examples of that by any chance?

**Comment by fhemberger** [comment]
@phillipj No black magic, all changes are openly available on GitHub:
- marked: https://github.com/Snyk/vulndb/tree/snapshots/master/patches/npm/marked
- minimatch: https://github.com/Snyk/vulndb/tree/snapshots/master/patches/npm/minimatch/20160620

**Comment by phillipj** [comment]
Aah, so snyk installs its own version of e.g. minimatch in prepublish - that's the blackmagic I was looking for.

Again snyk seems too intrusive and possibly misleading for my taste, but I won't hold it back if most of you like it, so -0 from me.

**Comment by NubleX** [comment]
i got 80% stability mark on my node [😣]

**Comment by fhemberger** [comment]
@phillipj Not exactly. It installs the official package from npm, then patches the vulnerability (if patching is possibe).

E.g: Dependency `x@latest` still uses package `y@2.0.1`, which contains a vulnerability. It is reported to the author and gets patched, `y@3.0.0`. Still, the author of `x` takes their time to update the dependency (or isn't even aware of the update). So snyk tries to backport the patch to `y@2.0.1` (if possible) for the time being. It's the same "black magic" Linux package maintainers have been doing for years. 😃 

I've been trying myself to reach out to npm package maintainers to address dependency vulnerability issues, it's very hard work and PRs are being ignored some times for ages even if you make it clear, that a dependency is vulnerable and a security issue. Especially if the package is n-levels deep in the dependency chain. So snyk tries to mitigate this problem.

**Comment by fhemberger** [comment]
I've added a clean version of this PR.

**Comment by NubleX** [comment]
nice one :)

**Comment by fhemberger** [comment]
Closing this for #841.
